pkg/wakaama/contrib/objects/on_off_switch.c

```diff
@@ -0,0 +1,492 @@
+                         lwm2m_data_t * data_array, lwm2m_object_t * object, lwm2m_write_type_t write_type);
+
+/**
+ * @brief Gets the current value of a given @p instance.
+ *
+ * @param[in, out] data     Initialized data structure.
+ * @param[in] instance      Pointer to the instance to get the value from.
+ *
+ * @return COAP_205_CONTENT on success
+ * @return COAP_404_NOT_FOUND if the value is not found
+ */
+static uint8_t _get_value(lwm2m_data_t *data, lwm2m_obj_on_off_switch_inst_t *instance);
+
+/**
+ * @brief Mark a resource as changed on the LwM2M engine.
+ *
+ * @param[in] instance_id    ID of the instance to mark the resource as changed.
+ * @param[in] resource_id    ID of the resource to mark as changed.
+ */
+static void _mark_resource_changed(uint16_t instance_id, uint16_t resource_id);
+
+struct lwm2m_on_off_switch_object {
+    lwm2m_object_t wakaama_object;                   /**< Wakaama internal object */
+    mutex_t lock;                                    /**< mutex for the instances access */
+    lwm2m_obj_on_off_switch_inst_t *free_instances;  /**< list of free instances */
+    lwm2m_obj_on_off_switch_inst_t instances[CONFIG_LWM2M_ON_OFF_SWITCH_INSTANCES_MAX]; /**< instances */
+};
+
+struct lwm2m_on_off_switch_object _on_off_switch_object = {
+    .lock = MUTEX_INIT,
+    .wakaama_object = {
+        .objID = LWM2M_ON_OFF_SWITCH_OBJECT_ID,
+        .instanceList = NULL,
+        .readFunc = _read_cb,
+        .writeFunc = _write_cb,
+        .executeFunc = NULL,
+        .createFunc = NULL,
+        .deleteFunc = NULL,
+        .discoverFunc = NULL,
+        .userData = NULL
+    }
+};
+
+static uint8_t _get_value(lwm2m_data_t *data, lwm2m_obj_on_off_switch_inst_t *instance)
+{
+    assert(data);
+    assert(instance);
+
+    switch (data->id) {
+    case LWM2M_ON_OFF_SWITCH_DIGITAL_INPUT_STATE_ID:
+        lwm2m_data_encode_bool(instance->status, data);
+        break;
+
+    case LWM2M_ON_OFF_SWITCH_ON_TIME_ID:
+        if (instance->status) {
+            int64_t time = (int64_t) ztimer_stopwatch_measure(&instance->stopwatch);
+            lwm2m_data_encode_int(time, data);
+        }
+        else {
+            lwm2m_data_encode_int(0, data);
+        }
+        break;
+
+    case LWM2M_ON_OFF_SWITCH_OFF_TIME_ID:
+        if (!instance->status) {
+            int64_t time = (int64_t) ztimer_stopwatch_measure(&instance->stopwatch);
+            lwm2m_data_encode_int(time, data);
+        }
+        else {
+            lwm2m_data_encode_int(0, data);
+        }
+        break;
+
+    case LWM2M_ON_OFF_SWITCH_APP_TYPE_ID:
+        lwm2m_data_encode_string(instance->app_type, data);
+        break;
+
+    default:
+        return COAP_404_NOT_FOUND;
+    }
+    return COAP_205_CONTENT;
+}
+
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
+                        lwm2m_data_t ** data_array, lwm2m_object_t * object)
+{
+    (void)context;
+
```

Comment:
```suggestion
static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
                        lwm2m_data_t ** p_data_array, lwm2m_object_t * object)
{
    (void)context;
    lwm2m_data_t * data_array = *p_data_array;

```

My idea is to get rid of the de-referencing in the rest of the code block. Especially breaking the `&(*data_array)[i]` down into `&data_array[i]` seems worth the effort for me.

Reply:
Changed